File: sdk/clojure/src/rocksky/library.clj
(ns rocksky.library
  "Ergonomic wrappers for the authenticated app.rocksky.library.* API (uploaded
  music). Build a client once with `client` and pass it to every call — the
  access token is bound in the client, not repeated:

    (def lib (library/client token))
    (library/get-genres lib)
    (library/get-song lib song-id)

  Optional params are keyword args merged into the request (camelCase keyword
  keys become the wire keys). Returns the parsed JSON payload."
  (:require [rocksky.core :as core]))

(defn client
  "Bind a (required, non-empty) access token; base overrides the AppView URL."
  ([token] (client token nil))
  ([token base] {::token token ::base base}))

(defn- lget [{::keys [token base]} nsid params]
  (core/library-get token nsid params base))

(defn- lpost [{::keys [token base]} nsid body]
  (core/library-post token nsid body base))

;; app.rocksky.library.ping
(defn ping [client]
  (lget client "app.rocksky.library.ping" {}))

;; app.rocksky.library.getLicense
(defn get-license [client]
  (lget client "app.rocksky.library.getLicense" {}))

;; app.rocksky.library.getMusicFolders
(defn get-music-folders [client]
  (lget client "app.rocksky.library.getMusicFolders" {}))

;; app.rocksky.library.getScanStatus
(defn get-scan-status [client]
  (lget client "app.rocksky.library.getScanStatus" {}))

;; app.rocksky.library.startScan
(defn start-scan [client]
  (lget client "app.rocksky.library.startScan" {}))

;; app.rocksky.library.getUser
(defn get-user [client]
  (lget client "app.rocksky.library.getUser" {}))

;; app.rocksky.library.getArtists
(defn get-artists [client]
  (lget client "app.rocksky.library.getArtists" {}))

;; app.rocksky.library.getIndexes
(defn get-indexes [client]
  (lget client "app.rocksky.library.getIndexes" {}))

;; app.rocksky.library.getArtist
(defn get-artist [client id]
  (lget client "app.rocksky.library.getArtist" {:id id}))

;; app.rocksky.library.getArtistInfo
(defn get-artist-info [client id]
  (lget client "app.rocksky.library.getArtistInfo" {:id id}))

;; app.rocksky.library.getAlbum
(defn get-album [client id]
  (lget client "app.rocksky.library.getAlbum" {:id id}))

;; app.rocksky.library.getAlbumList
(defn get-album-list [client type & {:as opts}]
  (lget client "app.rocksky.library.getAlbumList" (merge {:type type} opts)))

;; app.rocksky.library.getAlbumInfo
(defn get-album-info [client id]
  (lget client "app.rocksky.library.getAlbumInfo" {:id id}))

;; app.rocksky.library.getSong
(defn get-song [client id]
  (lget client "app.rocksky.library.getSong" {:id id}))

;; app.rocksky.library.getRandomSongs
(defn get-random-songs [client & {:as opts}]
  (lget client "app.rocksky.library.getRandomSongs" opts))

;; app.rocksky.library.getSongsByGenre
(defn get-songs-by-genre [client genre & {:as opts}]
  (lget client "app.rocksky.library.getSongsByGenre" (merge {:genre genre} opts)))

;; app.rocksky.library.getSimilarSongs
(defn get-similar-songs [client id & {:as opts}]
  (lget client "app.rocksky.library.getSimilarSongs" (merge {:id id} opts)))

;; app.rocksky.library.getTopSongs
(defn get-top-songs [client artist & {:as opts}]
  (lget client "app.rocksky.library.getTopSongs" (merge {:artist artist} opts)))

;; app.rocksky.library.getLyrics
(defn get-lyrics [client & {:as opts}]
  (lget client "app.rocksky.library.getLyrics" opts))

;; app.rocksky.library.getMusicDirectory
(defn get-music-directory [client id]
  (lget client "app.rocksky.library.getMusicDirectory" {:id id}))

;; app.rocksky.library.getGenres
(defn get-genres [client]
  (lget client "app.rocksky.library.getGenres" {}))

;; app.rocksky.library.search
(defn search [client query & {:as opts}]
  (lget client "app.rocksky.library.search" (merge {:query query} opts)))

;; app.rocksky.library.getStarred
(defn get-starred [client]
  (lget client "app.rocksky.library.getStarred" {}))

;; app.rocksky.library.star
(defn star [client id & {:as opts}]
  (lpost client "app.rocksky.library.star" (merge {:id id} opts)))

;; app.rocksky.library.unstar
(defn unstar [client id & {:as opts}]
  (lpost client "app.rocksky.library.unstar" (merge {:id id} opts)))

;; app.rocksky.library.getPlaylists
(defn get-playlists [client]
  (lget client "app.rocksky.library.getPlaylists" {}))

;; app.rocksky.library.getPlaylist
(defn get-playlist [client id]
  (lget client "app.rocksky.library.getPlaylist" {:id id}))

;; app.rocksky.library.createPlaylist
(defn create-playlist [client name]
  (lpost client "app.rocksky.library.createPlaylist" {:name name}))

;; app.rocksky.library.updatePlaylist
(defn update-playlist [client playlist-id & {:as opts}]
  (lpost client "app.rocksky.library.updatePlaylist" (merge {:playlistId playlist-id} opts)))

;; app.rocksky.library.deletePlaylist
(defn delete-playlist [client id]
  (lpost client "app.rocksky.library.deletePlaylist" {:id id}))

;; app.rocksky.library.deleteSong
(defn delete-song [client id]
  (lpost client "app.rocksky.library.deleteSong" {:id id}))

;; app.rocksky.library.deleteAlbum
(defn delete-album [client id]
  (lpost client "app.rocksky.library.deleteAlbum" {:id id}))

;; app.rocksky.library.scrobble
(defn scrobble [client id & {:as opts}]
  (lpost client "app.rocksky.library.scrobble" (merge {:id id} opts)))

;; app.rocksky.library.updateNowPlaying
(defn update-now-playing [client id]
  (lpost client "app.rocksky.library.updateNowPlaying" {:id id}))

;; app.rocksky.library.getNowPlaying
(defn get-now-playing [client]
  (lget client "app.rocksky.library.getNowPlaying" {}))

;; app.rocksky.library.getPlayQueue
(defn get-play-queue [client]
  (lget client "app.rocksky.library.getPlayQueue" {}))

;; app.rocksky.library.savePlayQueue
(defn save-play-queue [client & {:as opts}]
  (lpost client "app.rocksky.library.savePlayQueue" opts))

;; app.rocksky.library.getStreamUrl
(defn get-stream-url [client id & {:as opts}]
  (lget client "app.rocksky.library.getStreamUrl" (merge {:id id} opts)))

;; app.rocksky.library.getDownloadUrl
(defn get-download-url [client id]
  (lget client "app.rocksky.library.getDownloadUrl" {:id id}))

;; app.rocksky.library.getCoverArtUrl
(defn get-cover-art-url [client id & {:as opts}]
  (lget client "app.rocksky.library.getCoverArtUrl" (merge {:id id} opts)))

;; app.rocksky.library.getInternetRadioStations
(defn get-internet-radio-stations [client]
  (lget client "app.rocksky.library.getInternetRadioStations" {}))

File: sdk/ruby/lib/rocksky/library.rb
# frozen_string_literal: true

module Rocksky
  # Authenticated client for the app.rocksky.library.* API — the Subsonic /
  # navidrome-compatible surface over a user's uploaded music. Every method
  # requires auth, so a token is mandatory at construction (see Rocksky.library).
  # Optional params are Ruby keyword args (nil = omitted); wire keys stay
  # camelCase. Returns the parsed JSON payload.
  class Library
    def initialize(token, base: nil)
      raise Error, "app.rocksky.library.* requires an access token" if token.nil? || token.to_s.empty?

      @token = token.to_s
      @base = base.to_s
    end

    # app.rocksky.library.ping
    def ping()
      query("app.rocksky.library.ping", {})
    end

    # app.rocksky.library.getLicense
    def get_license()
      query("app.rocksky.library.getLicense", {})
    end

    # app.rocksky.library.getMusicFolders
    def get_music_folders()
      query("app.rocksky.library.getMusicFolders", {})
    end

    # app.rocksky.library.getScanStatus
    def get_scan_status()
      query("app.rocksky.library.getScanStatus", {})
    end

    # app.rocksky.library.startScan
    def start_scan()
      query("app.rocksky.library.startScan", {})
    end

    # app.rocksky.library.getUser
    def get_user()
      query("app.rocksky.library.getUser", {})
    end

    # app.rocksky.library.getArtists
    def get_artists()
      query("app.rocksky.library.getArtists", {})
    end

    # app.rocksky.library.getIndexes
    def get_indexes()
      query("app.rocksky.library.getIndexes", {})
    end

    # app.rocksky.library.getArtist
    def get_artist(id)
      query("app.rocksky.library.getArtist", { id: id })
    end

    # app.rocksky.library.getArtistInfo
    def get_artist_info(id)
      query("app.rocksky.library.getArtistInfo", { id: id })
    end

    # app.rocksky.library.getAlbum
    def get_album(id)
      query("app.rocksky.library.getAlbum", { id: id })
    end

    # app.rocksky.library.getAlbumList
    def get_album_list(type, size: nil, offset: nil, from_year: nil, to_year: nil, genre: nil)
      query("app.rocksky.library.getAlbumList", { type: type, size: size, offset: offset, fromYear: from_year, toYear: to_year, genre: genre })
    end

    # app.rocksky.library.getAlbumInfo
    def get_album_info(id)
      query("app.rocksky.library.getAlbumInfo", { id: id })
    end

    # app.rocksky.library.getSong
    def get_song(id)
      query("app.rocksky.library.getSong", { id: id })
    end

    # app.rocksky.library.getRandomSongs
    def get_random_songs(size: nil, genre: nil, from_year: nil, to_year: nil)
      query("app.rocksky.library.getRandomSongs", { size: size, genre: genre, fromYear: from_year, toYear: to_year })
    end

    # app.rocksky.library.getSongsByGenre
    def get_songs_by_genre(genre, count: nil, offset: nil)
      query("app.rocksky.library.getSongsByGenre", { genre: genre, count: count, offset: offset })
    end

    # app.rocksky.library.getSimilarSongs
    def get_similar_songs(id, count: nil)
      query("app.rocksky.library.getSimilarSongs", { id: id, count: count })
    end

    # app.rocksky.library.getTopSongs
    def get_top_songs(artist, count: nil)
      query("app.rocksky.library.getTopSongs", { artist: artist, count: count })
    end

    # app.rocksky.library.getLyrics
    def get_lyrics(artist: nil, title: nil)
      query("app.rocksky.library.getLyrics", { artist: artist, title: title })
    end

    # app.rocksky.library.getMusicDirectory
    def get_music_directory(id)
      query("app.rocksky.library.getMusicDirectory", { id: id })
    end

    # app.rocksky.library.getGenres
    def get_genres()
      query("app.rocksky.library.getGenres", {})
    end

    # app.rocksky.library.search
    def search(query, artist_count: nil, artist_offset: nil, album_count: nil, album_offset: nil, song_count: nil, song_offset: nil)
      query("app.rocksky.library.search", { query: query, artistCount: artist_count, artistOffset: artist_offset, albumCount: album_count, albumOffset: album_offset, songCount: song_count, songOffset: song_offset })
    end

    # app.rocksky.library.getStarred
    def get_starred()
      query("app.rocksky.library.getStarred", {})
    end

    # app.rocksky.library.star
    def star(id, album_id: nil, artist_id: nil)
      procedure("app.rocksky.library.star", { id: id, albumId: album_id, artistId: artist_id })
    end

    # app.rocksky.library.unstar
    def unstar(id, album_id: nil, artist_id: nil)
      procedure("app.rocksky.library.unstar", { id: id, albumId: album_id, artistId: artist_id })
    end

    # app.rocksky.library.getPlaylists
    def get_playlists()
      query("app.rocksky.library.getPlaylists", {})
    end

    # app.rocksky.library.getPlaylist
    def get_playlist(id)
      query("app.rocksky.library.getPlaylist", { id: id })
    end

    # app.rocksky.library.createPlaylist
    def create_playlist(name)
      procedure("app.rocksky.library.createPlaylist", { name: name })
    end

    # app.rocksky.library.updatePlaylist
    def update_playlist(playlist_id, name: nil, comment: nil, song_id_to_add: nil, song_index_to_remove: nil)
      procedure("app.rocksky.library.updatePlaylist", { playlistId: playlist_id, name: name, comment: comment, songIdToAdd: song_id_to_add, songIndexToRemove: song_index_to_remove })
    end

    # app.rocksky.library.deletePlaylist
    def delete_playlist(id)
      procedure("app.rocksky.library.deletePlaylist", { id: id })
    end

    # app.rocksky.library.deleteSong
    def delete_song(id)
      procedure("app.rocksky.library.deleteSong", { id: id })
    end

    # app.rocksky.library.deleteAlbum
    def delete_album(id)
      procedure("app.rocksky.library.deleteAlbum", { id: id })
    end

    # app.rocksky.library.scrobble
    def scrobble(id, time: nil, submission: nil)
      procedure("app.rocksky.library.scrobble", { id: id, time: time, submission: submission })
    end

    # app.rocksky.library.updateNowPlaying
    def update_now_playing(id)
      procedure("app.rocksky.library.updateNowPlaying", { id: id })
    end

    # app.rocksky.library.getNowPlaying
    def get_now_playing()
      query("app.rocksky.library.getNowPlaying", {})
    end

    # app.rocksky.library.getPlayQueue
    def get_play_queue()
      query("app.rocksky.library.getPlayQueue", {})
    end

    # app.rocksky.library.savePlayQueue
    def save_play_queue(id: nil, current: nil, position: nil)
      procedure("app.rocksky.library.savePlayQueue", { id: id, current: current, position: position })
    end

    # app.rocksky.library.getStreamUrl
    def get_stream_url(id, max_bit_rate: nil, format: nil)
      query("app.rocksky.library.getStreamUrl", { id: id, maxBitRate: max_bit_rate, format: format })
    end

    # app.rocksky.library.getDownloadUrl
    def get_download_url(id)
      query("app.rocksky.library.getDownloadUrl", { id: id })
    end

    # app.rocksky.library.getCoverArtUrl
    def get_cover_art_url(id, size: nil)
      query("app.rocksky.library.getCoverArtUrl", { id: id, size: size })
    end

    # app.rocksky.library.getInternetRadioStations
    def get_internet_radio_stations()
      query("app.rocksky.library.getInternetRadioStations", {})
    end

    private

    def query(nsid, params = {})
      Rocksky.unwrap(C.rocksky_library_get(@base, @token, nsid, JSON.generate(params.compact)))
    end

    def procedure(nsid, body = {})
      Rocksky.unwrap(C.rocksky_library_post(@base, @token, nsid, JSON.generate(body.compact)))
    end
  end
end
